src/test/java/ReadingAggregatorTest.java

```diff
@@ -0,0 +1,125 @@
+import eventprocessing.fileservices.CSVFileService;
+import eventprocessing.models.Bucket;
+import eventprocessing.models.ReadingAggregator;
+import eventprocessing.models.SqsResponse;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+import org.threeten.extra.MutableClock;
+
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.temporal.ChronoUnit;
+
+import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+
+public class ReadingAggregatorTest {
+
+    private ReadingAggregator readingAggregator;
+    private CSVFileService fileService;
+    private MutableClock clock;
+
+    @Before
+    public void buildReadingAggregator() {
+        clock = buildClock();
+        fileService = Mockito.mock(CSVFileService.class);
+        readingAggregator = new ReadingAggregator(fileService, clock);
+    }
+
+    @Test
+    public void process_nothingWhileNoReadingsAreOlderThanDelayTime() {
+        SqsResponse reading = buildReading(0);
+        readingAggregator.process(reading);
+        verify(fileService, never()).write(any());
+    }
+
+    @Test
+    public void process_singleReadingExceedingDelayTime_isSentToFileWriter() {
+        SqsResponse oldReading = buildReading(-5);
+        SqsResponse newReading = buildReading(0);
+
+        readingAggregator.process(oldReading);
+        advanceClockByMinutes(1);
+        readingAggregator.process(newReading);
+
+        assertWriteCalledTimesWithReadings(1, oldReading);
+    }
+
+    @Test
+    public void process_multipleReadingsExceedingDelayTime_areSentToFileWriter() {
+        SqsResponse oldReading1 = buildReading(-5);
+        SqsResponse oldReading2 = buildReading(-4.9);
+        SqsResponse oldReading3 = buildReading(-4);
+        SqsResponse newReading = buildReading(0);
+
+        processMultipleReadings(oldReading1, oldReading2, oldReading3);
+        advanceClockByMinutes(1);
+        readingAggregator.process(newReading);
+
+        assertWriteCalledTimesWithReadings(1, oldReading1, oldReading2);
+    }
+
+    @Test
+    public void process_sendsEachReadingToFileWriterOnlyOnce() {
+        SqsResponse oldReading1 = buildReading(-5);
+        SqsResponse oldReading2 = buildReading(-4.9);
+        SqsResponse oldReading3 = buildReading(-4);
+        SqsResponse newReading = buildReading(0);
+
+        processMultipleReadings(oldReading1, oldReading2, oldReading3);
+        advanceClockByMinutes(1);
+        readingAggregator.process(newReading);
+        advanceClockByMinutes(1);
+        readingAggregator.process(newReading);
+
+        assertWriteCalledTimesWithReadings(2, oldReading3);
```

Comment:
This assertion is slightly confusing. When I first read it, I thought it was checking that `oldReading3` is written to the file _twice_. But really it means that `csvWriter` is called twice, and the last call just contains `oldReading3`.

One option would be have an assertion that takes multiple expected sets of readings, and uses the `ArgumentCaptor` `getAllValues` method to check that all the buckets were correct. But have a think in case you can come up with a better assertion.